python/hsfs/constructor/query.py

```diff
@@ -50,6 +69,95 @@ def __init__(
         self._storage_connector_api = storage_connector_api.StorageConnectorApi(
             feature_store_id
         )
+        if self._left_feature_group is not None and self._left_features is not None:
+            self._populate_collections()
+
+    def _add_to_collection(self, feat, prefix, featuregroup, query_feature=True):
+        collection = (
+            self._query_features if query_feature else self._featuregroup_features
+        )
+        feature_entry = (feat, prefix, featuregroup)
+        collection[feat.name] = collection.get(feat.name, []) + [feature_entry]
+        if query_feature:
+            if prefix:
+                name_with_prefix = f"{prefix}{feat.name}"
+                collection[name_with_prefix] = collection.get(name_with_prefix, []) + [
+                    feature_entry
+                ]
+            self._query_feature_list.append(feature_entry)
+
+    def _populate_collections(self):
+        self._featuregroups = {self._left_feature_group}
+        self._query_features = {}
+        self._query_feature_list = []
+        self._featuregroup_features = {}
+        self._filters = self._filter
```

Comment:
I'm not sure if you wean that, but we (re-)initialize the lists/dicts here so they can be filled from scratch? We maintain a list and a map of features, to cater to different access patterns. (list for `.features` and dict for `get_feature` etc).

Reply:
oh no here I meant specifically this line: `self._filters = self._filter`

Reply:
Also, I get why you want all of these variables. But the problem (at least to me) is that it is hard to follow what goes into each one and why. I believe this will cause future issues, not only consistency (what you expect to be on the map might not be there even though it is in the list) but also even minor changes will require potentially a lot of debugging (for example someone decided to change the order of tuple, etc.).

Reply:
Regarding the filters - every join can contain an additional filter and I collect them all across the joins, starting with the filter in the root query object

Reply:
I guess the hard to follow part is also partly the fliside of reducing code duplication :) 
I'm open to suggestions on how to make it easier to follow, but I don't share the opinion that it the double bookkeeping or tuple access by index will inevitably lead to issues in the future :) That's also what we have code review for

Reply:
@bubriks ended up making the change. Can you have a second look?